src/xrpld/app/misc/CredentialHelpers.cpp

```diff
@@ -0,0 +1,272 @@
+        {
+            JLOG(j.fatal()) << "Unable to delete Credential from owner.";
+            return tefBAD_LEDGER;
+        }
+
+        if (isOwner)
+            adjustOwnerCount(view, sleAccount, -1, j);
+
+        return tesSUCCESS;
+    };
+
+    auto const issuer = sleCredential->getAccountID(sfIssuer);
+    auto const subject = sleCredential->getAccountID(sfSubject);
+    bool const accepted = sleCredential->getFlags() & lsfAccepted;
+
+    auto err = delSLE(issuer, sfIssuerNode, !accepted || (subject == issuer));
+    if (!isTesSuccess(err))
+        return err;
+
+    if (subject != issuer)
+    {
+        err = delSLE(subject, sfSubjectNode, accepted);
+        if (!isTesSuccess(err))
+            return err;
+    }
+
+    // Remove object from ledger
+    view.erase(sleCredential);
+
+    return tesSUCCESS;
+}
+
+NotTEC
+checkFields(PreflightContext const& ctx)
+{
+    if (!ctx.tx.isFieldPresent(sfCredentialIDs))
+        return tesSUCCESS;
+
+    auto const& credentials = ctx.tx.getFieldV256(sfCredentialIDs);
+    if (credentials.empty() || (credentials.size() > maxCredentialsArraySize))
+    {
+        JLOG(ctx.j.trace())
+            << "Malformed transaction: Credentials array size is invalid: "
+            << credentials.size();
+        return temMALFORMED;
+    }
+
+    std::unordered_set<uint256> duplicates;
+    for (auto const& cred : credentials)
+    {
+        auto [it, ins] = duplicates.insert(cred);
+        if (!ins)
+        {
+            JLOG(ctx.j.trace())
+                << "Malformed transaction: duplicates in credentials.";
+            return temMALFORMED;
+        }
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+valid(PreclaimContext const& ctx, AccountID const& src)
+{
+    if (!ctx.tx.isFieldPresent(sfCredentialIDs))
+        return tesSUCCESS;
+
+    auto const& credIDs(ctx.tx.getFieldV256(sfCredentialIDs));
+    for (auto const& h : credIDs)
+    {
+        auto const sleCred = ctx.view.read(keylet::credential(h));
+        if (!sleCred)
+        {
+            JLOG(ctx.j.trace()) << "Credential doesn't exist. Cred: " << h;
+            return tecBAD_CREDENTIALS;
+        }
+
+        if (sleCred->getAccountID(sfSubject) != src)
+        {
+            JLOG(ctx.j.trace())
+                << "Credential doesn’t belong to the source account. Cred: "
+                << h;
+            return tecBAD_CREDENTIALS;
+        }
+
+        if (!(sleCred->getFlags() & lsfAccepted))
+        {
+            JLOG(ctx.j.trace()) << "Credential isn't accepted. Cred: " << h;
+            return tecBAD_CREDENTIALS;
+        }
+
+        // Expiration checks are in doApply
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+authorized(ApplyContext const& ctx, AccountID const& dst)
+{
+    auto const& credIDs(ctx.tx.getFieldV256(sfCredentialIDs));
+    std::set<std::pair<AccountID, Slice>> sorted;
+    std::vector<std::shared_ptr<SLE const>> lifeExtender;
+    lifeExtender.reserve(credIDs.size());
```

Comment:
Is this needed because of `Slice`? Wouldn't it be cached? Can make it further efficient and calculate the hash here instead of iterating again in `depositPreauth`. Just need to add `depositPreauth` overload, which takes the hash.

Reply:
nvm, it has to be sorted. still, this lifeExtender feels a bit odd.

Reply:
>this lifeExtender feels a bit odd.
between moving shared_ptr and  or copying  credentials string I choose  move.